### PR TITLE
[Ref][Core][Opset13] NMSRotated-13 core shell and reference implementation

### PR DESCRIPTION
**Pull request by mitruska**
### Details:
 - NMSRotated-13 core shell and reference implementation
 - The base code and logic for v9 non_max_suppresion reference is almost the same (potential unification can be applied later), the main differences are:
 -- box representation as rotated box (x_ctr, y_ctr, width, height, angle_radians)
 -- calculation of intersection over union for rotated boxes (area of convex polygon needed)
 -- clockwise attribute for rotation direction
 - PR with spec: https://github.com/openvinotoolkit/openvino/pull/19574
 - ONNX FE NMSRotated import (preview)

### Tickets:
 - 115301, 115306, 115302, 118945

### CODE REVIEW

File: src/core/include/openvino/opsets/opset13_tbl.hpp
Comment (mitruska) [reply src/core/include/openvino/opsets/opset13_tbl.hpp:10]: Seems to be possible, as long as there are only new ops in Opset13 (no new version any of the existing operation).

File: src/plugins/template/tests/functional/op_reference/nms_rotated.cpp
Comment (t-jankowski) [inline src/plugins/template/tests/functional/op_reference/nms_rotated.cpp:74]:
```suggestion
    static std::shared_ptr<Model> CreateModel(const NMSRotatedParams& params) {
```

File: src/plugins/template/tests/functional/op_reference/nms_rotated.cpp
Comment (t-jankowski) [inline src/plugins/template/tests/functional/op_reference/nms_rotated.cpp:104]:
```suggestion
        model = CreateModel(params);
```

**Review by t-jankowski — CHANGES_REQUESTED** [review]

File: src/core/reference/include/openvino/reference/nms_rotated_util.hpp
Comment (praasz) [inline src/core/reference/include/openvino/reference/nms_rotated_util.hpp:68]: General comment function make some comment what they are doing. It could give better overview what is intention for implementation.
I'm not sure what exactly means get intersection points? But there can be intersection point of two lines.

File: src/core/reference/include/openvino/reference/nms_rotated_util.hpp
Comment (praasz) [inline src/core/reference/include/openvino/reference/nms_rotated_util.hpp:144]:
```suggestion
                                     const int num_in,
```

File: src/core/src/op/nms_rotated.cpp
Comment (mitruska) [reply src/core/src/op/nms_rotated.cpp:21]: 5 shapes can be provided to `shape_infer`, without any input connected to the op, so I think it's worth to check the input size as well.

File: src/core/src/op/nms_rotated.cpp
Comment (mitruska) [reply src/core/src/op/nms_rotated.cpp:51]: Is there anything wrong with usage of `op::v13::` to define functions in *.cpp file?
Also if there is more than one op version in the file, the version can be seen explicitly.
This is the convention used in most of the op *.cpp files now.

https://github.com/openvinotoolkit/openvino/blob/26d18c924b36bfaec169dffc85d5b35c58b666e7/src/core/src/op/scatter_elements_update.cpp#L323-L328

https://github.com/openvinotoolkit/openvino/blob/26d18c924b36bfaec169dffc85d5b35c58b666e7/src/core/src/op/scatter_elements_update.cpp#L15-L24

File: src/core/tests/type_prop/nms_rotated.cpp
Comment (mitruska) [reply src/core/tests/type_prop/nms_rotated.cpp:20]: It's not so trivial as there is a difference in expected shape of the boxes input, and "default" inputs.
So the intention of usage of TYPED_TEST was indeed to reuse the NMS tests, anyway it can be useful for further version of the NMSRotated.

File: src/plugins/template/tests/functional/op_reference/nms_rotated.cpp
Comment (mitruska) [reply src/plugins/template/tests/functional/op_reference/nms_rotated.cpp:104]: The "function" is a member of the `CommonReferenceTest` class:

https://github.com/openvinotoolkit/openvino/blob/5611f7d65c4ee0b6a78ed460d0ab093c7f559321/src/plugins/template/tests/functional/op_reference/base_reference_test.hpp#L35

File: src/plugins/template/backend/ops/nms_rotated.cpp
Comment (mitruska) [reply src/plugins/template/backend/ops/nms_rotated.cpp:12]: The last `evaluate_node` need to be outside of ov namespace, the functions above are used just within this file.

File: src/frontends/onnx/frontend/src/op/nms_rotated.hpp
Comment (mitruska) [reply src/frontends/onnx/frontend/src/op/nms_rotated.hpp]: Not needed, removed.

File: src/plugins/template/tests/functional/op_reference/nms_rotated.cpp
Comment (t-jankowski) [reply src/plugins/template/tests/functional/op_reference/nms_rotated.cpp]: I mixed up the classes
https://github.com/openvinotoolkit/openvino/pull/19366/files#diff-811d4ce9d2f38671aadc9df3cac63c93f37b3a8d54c9a089446a90f154040293R34

File: src/core/shape_inference/include/nms_shape_inference.hpp
Comment (mitruska) [reply src/core/shape_inference/include/nms_shape_inference.hpp:308]: To allow GPU implementation reuse the same mechanisms as for NMS, it can be better to keep all of the related `shape_infer` functions in sync. And it can be updated for all nms ops at once when redundant.

File: src/core/reference/include/openvino/reference/nms_rotated_util.hpp
Comment (mitruska) [reply src/core/reference/include/openvino/reference/nms_rotated_util.hpp:58]: In fact there is different calculation performed for each point coordinate.

File: src/core/reference/include/openvino/reference/nms_rotated_util.hpp
Comment (mitruska) [reply src/core/reference/include/openvino/reference/nms_rotated_util.hpp:54]: It's just a replacement of division by 2.

File: src/core/reference/src/op/nms_rotated.cpp
Comment (mitruska) [reply src/core/reference/src/op/nms_rotated.cpp:45]: The constructors are currently used to fill output vectors of different types.
But noted as a suggestion for common refactor of nms reference.

File: src/core/src/op/nms_rotated.cpp
Comment (mitruska) [reply src/core/src/op/nms_rotated.cpp:49]: Keeping aligned with the rest of the op .cpp files

File: src/core/src/op/nms_rotated.cpp
Comment (mitruska) [reply src/core/src/op/nms_rotated.cpp:72]: Keeping aligned with the rest of the op .cpp files

File: src/core/reference/include/openvino/reference/nms_rotated_util.hpp
Comment (mitruska) [reply src/core/reference/include/openvino/reference/nms_rotated_util.hpp:101]: Noted as possible future refactor

File: src/core/reference/include/openvino/reference/nms_rotated_util.hpp
Comment (mitruska) [reply src/core/reference/include/openvino/reference/nms_rotated_util.hpp:71]: Noted as possible future refactor

File: src/core/reference/include/openvino/reference/nms_rotated_util.hpp
Comment (mitruska) [reply src/core/reference/include/openvino/reference/nms_rotated_util.hpp:1]: Additional refactoring will be planned along with some possibilities of unification with current `non_max_suppression` reference implementation.

File: src/core/src/op/nms_rotated.cpp
Comment (mitruska) [reply src/core/src/op/nms_rotated.cpp]: Error message updated.

File: src/core/reference/include/openvino/reference/nms_rotated_util.hpp
Comment (mitruska) [reply src/core/reference/include/openvino/reference/nms_rotated_util.hpp]: Added more comments, the intersection points are the points defining boxes intersection area (intersection points of the edges and the corners within the boxes area). General algorithm can be also found in the spec [NMSRotated_13.md](https://github.com/openvinotoolkit/openvino/blob/master/docs/ops/sort/NMSRotated_13.md).

File: src/plugins/template/backend/ops/nms_rotated.cpp
Comment (mitruska) [reply src/plugins/template/backend/ops/nms_rotated.cpp]: Added comment with explanation here.

File: src/core/src/op/nms_rotated.cpp
Comment (mitruska) [reply src/core/src/op/nms_rotated.cpp:43]: Keeping aligned with the rest of the operators .cpp files.

File: src/core/reference/src/op/nms_rotated.cpp
Comment (mitruska) [reply src/core/reference/src/op/nms_rotated.cpp]: Lack of this `static` is triggering warnings.

File: src/core/src/op/nms_rotated.cpp
Comment (praasz) [reply src/core/src/op/nms_rotated.cpp:43]: Is not wrong but is less typing and there is the whole code block is in v13 namespace.
Is just detail.

File: src/core/src/op/nms_rotated.cpp
Comment (praasz) [reply src/core/src/op/nms_rotated.cpp:21]: Shape inference is done first and the same check is here again.
Or can be done like shape_infer is first (as it always check number of inputs connected to as shapes are taken from inputs) and then is not required to check it again.
Is just redundant.

File: src/core/shape_inference/include/nms_shape_inference.hpp
Comment (praasz) [reply src/core/shape_inference/include/nms_shape_inference.hpp:303]: Yes it should be with sync with common aprocha in shape inference and be compatible with `ShapeInferTA` and `ShapeInferBase`, there are few more types.
The GPU for NMS is like exception and should be kept as exception only. 

Introducing new signature can have consequence as another template has to implemented in shape inference instead using existing one (add new code increase bin size).
This flag is only required in GPU for one version and is was introduced before unification of `shape_infer` functions. The shape inference can force plugin to use its custom implementation on plugin side. The exception should be refactored to be aligned with common approach in shape inference (it was not done because of time). Is better to not introduce new way of shape inference and is better to remove this parameter.

File: src/core/reference/src/op/nms_rotated.cpp
Comment (praasz) [reply src/core/reference/src/op/nms_rotated.cpp:18]: there are warnings if function is in anonymous namespace?

File: src/core/shape_inference/include/nms_shape_inference.hpp
Comment (praasz) [inline src/core/shape_inference/include/nms_shape_inference.hpp:68]:
```suggestion
    TDim box_def_size = ov::is_type<v13::NMSRotated>(op) ? 5 : 4;
```
Detail

File: src/core/src/op/nms_rotated.cpp
Comment (praasz) [inline src/core/src/op/nms_rotated.cpp:25]: Detail: Here instead `continue` the check for integral input can be placed.

**Review by t-jankowski — APPROVED** [review]

File: src/core/reference/src/op/nms_rotated.cpp
Comment (mitruska) [reply src/core/reference/src/op/nms_rotated.cpp:18]: Yes, that the function has no declaration:
```
/reference/src/op/nms_rotated.cpp:18:7: warning: no previous declaration for ‘float ov::reference::nms_detail::rotated_intersection_over_union(const ov::reference::iou_rotated::RotatedBox&, const ov::reference::iou_rotated::RotatedBox&)’ [-Wmissing-declarations]
   18 | float rotated_intersection_over_union(const RotatedBox& boxI, const RotatedBox& boxJ) {
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: src/core/src/op/nms_rotated.cpp
Comment (mitruska) [reply src/core/src/op/nms_rotated.cpp:21]: The five inputs are required from the agreed NMSRotated specification perspective, but in fact the shapes of the last two inputs are not required to infer the output shape (but can be passed to validate correctness of the shapes). So the check in `shape_infer` is unified for nms and relaxed.

File: src/core/shape_inference/include/nms_shape_inference.hpp
Comment (mitruska) [reply src/core/shape_inference/include/nms_shape_inference.hpp]: Ok, as discussed `static_output` has been removed from the function parameters and resolved by local variable.

File: src/core/src/op/nms_rotated.cpp
Comment (praasz) [reply src/core/src/op/nms_rotated.cpp:21]: Ok, just missed that five inputs are mandatory.

**Review by praasz — APPROVED** [review]
Fix CIs, please.

**Comment by mitruska** [comment]
Waiting for legal approval (cc: @mlukasze)

File: src/core/tests/opset.cpp
Comment (praasz) [inline src/core/tests/opset.cpp:74]: Why is 180, not 179?

Comment (mitruska) [reply src/core/tests/opset.cpp:74]: 180 is the previous number +1, `BitwiseNot` op has been merged earlier

File: src/core/reference/include/openvino/reference/nms_rotated_util.hpp
Comment (mitruska) [reply src/core/reference/include/openvino/reference/nms_rotated_util.hpp:1]: Ticket for follow up refactoring: 122119

**Pull request merged**
